vcf.c

```diff
@@ -1591,53 +1766,63 @@ bcf_hdr_t *vcf_hdr_read(htsFile *fp)
     if ( bcf_hdr_parse(h, txt.s) < 0 ) goto error;
 
     // check tabix index, are all contigs listed in the header? add the missing ones
-    tbx_t *idx = tbx_index_load(fp->fn);
+    idx = tbx_index_load(fp->fn);
     if ( idx )
     {
         int i, n, need_sync = 0;
-        const char **names = tbx_seqnames(idx, &n);
+        names = tbx_seqnames(idx, &n);
+        if (!names) goto error;
         for (i=0; i<n; i++)
         {
             bcf_hrec_t *hrec = bcf_hdr_get_hrec(h, BCF_HL_CTG, "ID", (char*) names[i], NULL);
             if ( hrec ) continue;
-            hrec = (bcf_hrec_t*) calloc(1,sizeof(bcf_hrec_t));
-            hrec->key = strdup("contig");
-            bcf_hrec_add_key(hrec, "ID", strlen("ID"));
-            bcf_hrec_set_val(hrec, hrec->nkeys-1, (char*) names[i], strlen(names[i]), 0);
-            bcf_hdr_add_hrec(h, hrec);
+            if (add_missing_contig_hrec(h, names[i]) < 0) goto error;
             need_sync = 1;
         }
+        if ( need_sync ) {
+            if (bcf_hdr_sync(h) < 0) goto error;
+        }
         free(names);
         tbx_destroy(idx);
-        if ( need_sync )
-            bcf_hdr_sync(h);
     }
     free(txt.s);
     return h;
 
  error:
+    if (idx) tbx_destroy(idx);
+    free(names);
     free(txt.s);
     if (h) bcf_hdr_destroy(h);
     return NULL;
 }
 
 int bcf_hdr_set(bcf_hdr_t *hdr, const char *fname)
 {
-    int i, n;
+    int i = 0, n = 0, save_errno;
     char **lines = hts_readlines(fname, &n);
     if ( !lines ) return 1;
     for (i=0; i<n-1; i++)
     {
         int k;
         bcf_hrec_t *hrec = bcf_hdr_parse_line(hdr,lines[i],&k);
-        if ( hrec ) bcf_hdr_add_hrec(hdr, hrec);
+        if (!hrec) goto fail;
+        if (bcf_hdr_add_hrec(hdr, hrec) < 0) goto fail;
         free(lines[i]);
+        lines[i] = NULL;
     }
-    bcf_hdr_parse_sample_line(hdr,lines[n-1]);
+    if (bcf_hdr_parse_sample_line(hdr, lines[n-1]) < 0) goto fail;
+    if (bcf_hdr_sync(hdr) < 0) goto fail;
     free(lines[n-1]);
     free(lines);
-    bcf_hdr_sync(hdr);
     return 0;
+
+ fail:
+    save_errno = errno;
```

Comment:
I'm not convinced free can do anything to errno.  If it does, it's not documented and although this doesn't forbid it from being changed, the same could be said of pretty much every single function.

Reply:
The same **is** said of [pretty much every single function](http://pubs.opengroup.org/onlinepubs/9699919799/functions/errno.html), which is why functions like `bcf_hdr_set()` that want to return a meaningful `errno` need to do this errno saving dance in their failure cleanup.

For `free()`, this is [POSIX Defect 385](http://austingroupbugs.net/view.php?id=385); if POSIX Issue 8 ever gets released :smile:, you can be sure that successful `free()` does not modify `errno` on compliant platforms. In the meantime, you probably can't be sure and I guess Rob has followed the super-paranoid example of some of my _hfile.c_ code…